accepted/2020/platform-exclusion/platform-exclusion.md

```diff
@@ -0,0 +1,217 @@
+                    AttributeTargets.Class |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Event |
+                    AttributeTargets.Method |
+                    AttributeTargets.Module |
+                    AttributeTargets.Property |
+                    AttributeTargets.Struct,
+                    AllowMultiple=true, Inherited=false)]
+    public sealed class UnsupportedOSPlatformAttribute : OSPlatformAttribute
+    {
+        public UnsupportedOSPlatformAttribute(string platformName);
+    }
+```
+
+The semantics of this attribute are as follows:
+
+* Marks an assembly, module, or API as unsupported on a given OS platform.
+* The `platformName` can contain no version number, a single version number, or
+  a range:
+    - `browser`. Marks the API as unsupported on all version of `browser`. This
+      will be used for APIs that can't be supported on a particular OS platform.
+    - `ios14.0`. Marks the API as unsupported starting with iOS 14.0. This will
+      be used for OS platform APIs that were removed/disabled.
+    - `windows0.0-10.0.19041`. Marks the API as unsupported for 10.0.19041 and
+      earlier. This will be used for APIs that were unsupported before but can
+      be supported starting with a later version of the OS. This is different
+      from applying `[MinimumOSPlatform("windows10.0.19041")]` in that the
+      presence of this attribute doesn't cause the API to be considered
+      unsupported on other platforms.
+
+### Analyzer behavior
+
+The existing analyzer that handles `[MinimumOSPlatform]` will be modified to
+handle `[UnsupportedOSPlatform]` (previously named `[RemovedInOSPlatform]`)
+with the semantics listed above.
+
+Let's look at a few examples of annotations what the corresponding guard clauses
+are that would prevent the analyzer from flagging usage of the API.
+
+```C#
+// The API is supported everywhere except when running in a browser
+[UnsupportedOSPlatform("browser")]
+public extern void Api();
+
+public void Api_Usage()
+{
+    if (!RuntimeInformation.IsOSPlatform("browser"))
+    {
+        Api();
+    }
+}
+```
+
+```C#
+// On Windows, the API was unsupported up to and including version 10.0.19041.
+// The API is considered supported everywhere else without constraints.
+[UnsupportedOSPlatform("windows0.0-10.0.19041")]
+public extern void Api();
+
+public void Api_Usage()
+{
+    if (!RuntimeInformation.IsOSPlatformOrEarlier("windows10.0.19041"))
+    {
+        Api();
+    }
+}
+```
+
+```C#
+// The API is only supported on iOS. There, it started to be supported in
+// version 12.0 and stopped being supported in version 14.
+[MinimumOSPlatform("iOS12.0")]
+[UnsupportedOSPlatform("ios14.0")]
+public extern void Api();
+
+public void Api_Usage()
+{
+    if (RuntimeInformation.IsOSPlatformOrLater("iOS12.0") &&
+        RuntimeInformation.IsOSPlatformOrEarlier("ios14.0"))
+    {
+        Api();
+    }
+}
+```
```

Comment:
I'm not in love with the string based approach either, but it seemed simple enough to give it a shot. FWIW, I'm not concerned about syntax issues because the entire design is around having an analyzer. We can trivially verify that the user entered sensible information.

I'm not fan of the enum and the default attribute. It makes the simple cases extremely verbose and error prone. If I want to mark an API as unsupported on a particular OS, I have to do this:

```
[OSPlatformSupport("ios", OSPlatformSupportLevel.Unupported)]
```

Then, when we make it supported we all of the sudden have to change this to:

```
[OSPlatformSupportDefault(OSPlatformSupportLevel.Supported)]
[OSPlatformSupport("ios12.0")]
```

That's not very intuitive. And we'd have another way to obsolete an API:

```
[OSPlatformSupportDefault(OSPlatformSupportLevel.Obsolete)]
```

I'm not saying my proposal is free of problems, but this particular spin looks worse to me than strings.

Reply:
> I'm not fan of the default attribute.

I'm honestly not thrilled with `OSPlatformDefault` either, but that concept seems to be what we're missing:
**When a platform is marked to be supported starting on a version, what does that mean for unspecified platforms? Are they _also_ supported (on all versions), or are they _unsupported_ entirely?**

> And we'd have another way to obsolete an API

We've already done that with `ObsoletedInOSPlatform`, right?  I was just trying to retain that concept here.  Should we revisit that?

Reply:
The new heuristic for sorting the platforms by version and identifying whether `SupportedOSPlatform` or `UnsupportedOSPlatform` comes first accomplishes the job of determining the default, without requiring any gymnastics that would be error-prone.  Resolving this thread.